### PR TITLE
Defined _repr_png_ for some modules

### PR DESCRIPTION
**Pull request by jinsooihm**
## Description

`_repr_png_` method for integration with `IPython` was added for some modules that have obvious representation as 2D diagrams.

- `cluster.Cluster`
- `density.ComplexCF`
- `density.FloatCF`
- `density.RDF`
- `environment.MatchEnv`
- `msd.MSD`
- `pmft.PMFTXY2D`
- `voronoi.Voronoi`

If there are more that need to be added, please comment.

## Motivation and Context

It looks nice on Jupyter

## How Has This Been Tested?

Tests have been added to see if the `_repr_png_` fails.
Also, if the input is 3D, some tests have been added to check if `_repr_png_` returns `None`.

## Types of changes

- [x] New feature (non-breaking change which adds or improves functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [ ] I have updated the [Changelog](https://github.com/glotzerlab/freud/blob/master/ChangeLog.md).

### CODE REVIEW

File: freud/cluster.pyx [hunk 1]
```diff
@@ -169,6 +169,24 @@ cdef class Cluster:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        try:
+            count = np.unique(self.cluster_idx, return_counts=True)
+        except ValueError:
+            print("computeClusters should be called before to show")
```
Comment (bdice) [inline freud/cluster.pyx]: Just remove this print statement - since `_repr_png_` is called implicitly by IPython, I think this may be confusing to the user.

File: freud/cluster.pyx [hunk 2]
```diff
@@ -169,6 +169,24 @@ cdef class Cluster:
...
+            return None
+        if not self.box.is2D():
+            return None
+        else:
+            count_sorted = sorted((freq, keys)
+                                  for keys, freq in zip(count[0], count[1]))
+            freqs = [i[0] for i in count_sorted[-10:]]
+            keys = [i[1] for i in count_sorted[-10:]]
```
Comment (bdice) [inline freud/cluster.pyx]: Is this going to return the clusters ordered like 📶 instead of largest to smallest? I'd prefer to use `sorted(..., key=lambda x: -x[0])` and `count_sorted[:10]` to show largest to smallest.

File: freud/common.py [hunk 3]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+
+    Args:
+        x (list): x values of the bar graph.
+        height (list): Height values corresponding to :code:`x`.
+        title (str): Title of the graph. (Default value = :code:`None`).
+        xlabel (str): Label of x axis. (Default value = :code:`None`).
+        ylabel (str): Label of y axis. (Default value = :code:`None`).
+
+    Returns:
+        bytes: Byte representation of the graph in png file if import
```
Comment (bdice) [inline freud/common.py]: `if matplotlib is available` is clearer than `if import succeeds`

File: freud/common.py [hunk 4]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        fig = Figure()
+        ax = fig.subplots()
+        ax.bar(x=x, height=height)
+        ax.set_title(title)
+        ax.set_xlabel(xlabel)
+        ax.set_ylabel(ylabel)
+        ax.set_xticks(x)
+        ax.set_xticklabels(x)
+        f = io.BytesIO()
+        canvas = FigureCanvasAgg(fig) # noqa F841
```
Comment (bdice) [inline freud/common.py]: What happens if you just write `FigureCanvasAgg(fig)` without the `canvas =`? Also please add a comment like `# Sets an Agg backend so this figure can be rendered` (or whatever you think makes the most sense).

File: freud/common.py [hunk 5]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        ylabel (str): Label of y axis. (Default value = :code:`None`).
+
+    Returns:
+        bytes: Byte representation of the graph in png file if import
+            succeeds. Otherwise :code:`None`.
+    """
+    try:
+        from matplotlib.figure import Figure
+        from matplotlib.backends.backend_agg import FigureCanvasAgg
+        import io
```
Comment (bdice) [inline freud/common.py]: We could put these module imports at the top.

```python
# Top of the file
import io
try:
    from matplotlib.figure import Figure
    from matplotlib.backends.backend_agg import FigureCanvasAgg
    MATPLOTLIB = True
except ImportError:
    MATPLOTLIB = False
```
...
```python
# in function definitions
if not MATPLOTLIB:
    return None
fig = Figure()
ax = fig.subplots()
```
...

File: freud/common.py [hunk 6]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        ax.set_title(title)
+        ax.set_xlabel(xlabel)
+        ax.set_ylabel(ylabel)
+        f = io.BytesIO()
+        canvas = FigureCanvasAgg(fig) # noqa F841
+        fig.savefig(f, format='png')
+        return f.getvalue()
+
+
+def make_polygon(sides, radius=1):
```
Comment (bdice) [inline freud/common.py]: I would remove this function since it's not used anywhere.

File: freud/common.py [hunk 7]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+                         for theta in thetas])
+    return vertices
+
+
+def pmft_plot(pmft):
+    """ Helper function to draw 2D PMFT diagram.
+
+    Args:
+        pmft (:class:`freud.pmft.PMFTXY2D`):
+            PMFT2DXY instance.
```
Comment (bdice) [inline freud/common.py]: Typo: PMFTXY2D

File: freud/common.py [hunk 8]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        import io
+        from scipy.ndimage.filters import gaussian_filter
+    except ImportError:
+        return None
+    else:
+        pmft_arr = np.copy(pmft.PMFT)
+
+        # Do some simple post-processing for plotting purposes
+        pmft_arr[np.isinf(pmft_arr)] = np.nan
+        dx = (2.0 * 3.0) / pmft.n_bins_X
```
Comment (bdice) [inline freud/common.py]: All this math used to set the bounding box is unfortunately not generalizable to all PMFTs. We should use the values in `pmft.X` and `pmft.Y`, because this example code is hard-coded for a specific shape size.

https://freud.readthedocs.io/en/stable/pmft.html#freud.pmft.PMFTXY2D

File: freud/common.py [hunk 9]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+            x = -3.0 + dx * i
+            for j in range(pmft.n_bins_Y):
+                y = -3.0 + dy * j
+                if ((x*x + y*y < 1.5) and (np.isnan(pmft_arr[j, i]))):
+                    pmft_arr[j, i] = 10.0
+        w = int(2.0 * pmft.n_bins_X / (2.0 * 3.0))
+        center = int(pmft.n_bins_X / 2)
+
+        # Get the center of the histogram bins
+        pmft_smooth = gaussian_filter(pmft_arr, 1)
```
Comment (bdice) [inline freud/common.py]: I'm not sure if smoothing is appropriate for displaying the raw data. Taking this out would also remove a dependency.

File: freud/common.py [hunk 10]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        reduced_y = y[center-w:center+w]
+
+        # Plot figures
+        fig = Figure(figsize=(12, 5), facecolor='white')
+        values = [-2, -1, 0, 2]
+        norm = Normalize(vmin=-2.5, vmax=3.0)
+        n_values = [norm(i) for i in values]
+        colors = viridis(n_values)
+        colors = colors[:, :3]
+        # verts = make_polygon(sides=6, radius=0.6204)
```
Comment (bdice) [inline freud/common.py]: Remove.

File: freud/common.py [hunk 11]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        x = pmft.X
+        y = pmft.Y
+        reduced_x = x[center-w:center+w]
+        reduced_y = y[center-w:center+w]
+
+        # Plot figures
+        fig = Figure(figsize=(12, 5), facecolor='white')
+        values = [-2, -1, 0, 2]
+        norm = Normalize(vmin=-2.5, vmax=3.0)
+        n_values = [norm(i) for i in values]
```
Comment (bdice) [inline freud/common.py]: We shouldn't hard-code these values in the previous few lines.

File: freud/common.py [hunk 12]
```diff
@@ -56,3 +56,246 @@ def convert_box(box):
...
+        fig.subplots_adjust(right=0.85)
+        cbar_ax = fig.add_axes([0.88, 0.1, 0.02, 0.8])
+        fig.colorbar(im, cax=cbar_ax)
+        f = io.BytesIO()
+        canvas = FigureCanvasAgg(fig) # noqa F841
+        fig.savefig(f, format='png')
+        return f.getvalue()
+
+
+def draw_voronoi(box, cells, color_by_sides=False):
```
Comment (bdice) [inline freud/common.py]: @klarh I was thinking about the related Voronoi-drawing code in `plato` - anything we should consider when creating this IPython representation of `freud` Voronoi objects?

File: freud/density.pyx [hunk 13]
```diff
@@ -211,6 +211,11 @@ cdef class FloatCF:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        return freud.common.line_plot(self.R, self.RDF,
+                                      title="Float CF",
```
Comment (bdice) [inline freud/density.pyx]: `title="Correlation Function"`

File: freud/density.pyx [hunk 14]
```diff
@@ -211,6 +211,11 @@ cdef class FloatCF:
...
+                                      xlabel="r", ylabel="p*q(r)")
```
Comment (bdice) [inline freud/density.pyx]: `ylabel="C(r)"`

File: freud/density.pyx [hunk 15]
```diff
@@ -402,6 +407,11 @@ cdef class ComplexCF:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        return freud.common.line_plot(self.R, np.real(self.RDF),
+                                      title="Complex CF",
```
Comment (bdice) [inline freud/density.pyx]: `title="Correlation Function"`

File: freud/density.pyx [hunk 16]
```diff
@@ -402,6 +407,11 @@ cdef class ComplexCF:
...
+                                      xlabel="r", ylabel="C(r)")
```
Comment (bdice) [inline freud/density.pyx]: Try this: `ylabel="\Re(C(r))"` - we want to indicate it's the Real part.

File: freud/environment.pyx [hunk 17]
```diff
@@ -869,6 +870,23 @@ cdef class MatchEnv:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        try:
+            count = np.unique(self.clusters, return_counts=True)
+        except ValueError:
+            print("clusters should be called before to show")
```
Comment (bdice) [inline freud/environment.pyx]: Remove as above.

File: freud/environment.pyx [hunk 18]
```diff
@@ -869,6 +870,23 @@ cdef class MatchEnv:
...
+            return None
+        if not self.m_box.is2D():
+            return None
+        count_sorted = sorted((freq, keys)
+                              for keys, freq in zip(count[0], count[1]))
+        freqs = [i[0] for i in count_sorted[-10:]]
+        keys = [i[1] for i in count_sorted[-10:]]
```
Comment (bdice) [inline freud/environment.pyx]: Re-sort as in `Cluster`.

File: tests/test_density_ComplexCF.py [hunk 19]
```diff
@@ -127,6 +127,23 @@ def test_repr(self):
...
+        num_points = 100
+        box_size = rmax*3.1
+        np.random.seed(0)
+        points = np.random.random_sample((num_points, 3)).astype(np.float32) \
+            * box_size - box_size/2
+        ang = np.random.random_sample((num_points)).astype(np.float64) \
+            * 2.0 * np.pi
+        comp = np.exp(1j*ang)
+        ocf = freud.density.ComplexCF(rmax, dr)
+        # first bin is bad
```
Comment (bdice) [inline tests/test_density_ComplexCF.py]: Is this comment just copy-pasted? I'm actually not sure what this refers to...

**Review by bdice — CHANGES_REQUESTED** [review]
Change requests below. Overall looks good. Can you post screenshots of the example plots? Just trying to make sure it looks realistic / useful.

File: freud/cluster.pyx (hunk 2 above)
Comment (jinsooihm) [reply freud/cluster.pyx]: Seems like you also need to make `keys` to be a list of string to make the bar plot be ordered from the largest cluster to the smallest.

File: tests/test_density_ComplexCF.py (hunk 19 above)
Comment (jinsooihm) [reply tests/test_density_ComplexCF.py]: It is copy-pasted.

File: freud/density.pyx (hunk 16 above)
Comment (jinsooihm) [reply freud/density.pyx]: Do people usually use this \Re for real part? I feel like ones I have seen just put normal text Re.

**Comment by jinsooihm** [comment]
Some screenshots:
- RDF
  Screen Shot 2019-06-10 at 1 34 26 PM
- Cluster / MatchEnv
  Screen Shot 2019-06-10 at 1 32 16 PM
- Complex CF
  Screen Shot 2019-06-10 at 1 41 07 PM
- Float CF
  Screen Shot 2019-06-10 at 1 42 07 PM
- Voronoi
  Screen Shot 2019-06-10 at 1 31 10 PM
- PMFTXY2D
  Screen Shot 2019-06-10 at 1 31 02 PM

**Comment by jinsooihm** [comment]
More screenshots:
- MSD window mode
  Screen Shot 2019-06-10 at 2 09 29 PM
- MSD direct mode
  Screen Shot 2019-06-10 at 2 09 25 PM

File: freud/common.py [hunk 20]
```diff
@@ -56,3 +63,266 @@ def convert_box(box):
...
+    Returns:
+        bytes: Byte representation of the graph in png file if matplotlib is
+        available. Otherwise :code:`None`.
+    """
+    count_sorted = sorted([(freq, key)
+                          for key, freq in zip(keys, freqs)],
+                          key=lambda x: -x[0])
+    freqs = [i[0] for i in count_sorted[:num_cluster_to_plot]]
+    keys = [str(i[1]) for i in count_sorted[:num_cluster_to_plot]]
+    return freud.common.bar_plot(keys, freqs,
```
Comment (bdice) [inline freud/common.py]: We should make it clear if there are more clusters than are being shown. Maybe add `(total clusters: N)` as a text annotation, or in the xlabel or title?

File: freud/density.pyx [hunk 21]
```diff
@@ -832,3 +844,7 @@ cdef class RDF:
 
     def __str__(self):
         return repr(self)
+
+    def _repr_png_(self):
+        return freud.common.line_plot(self.R, self.RDF,
```
Comment (bdice) [inline freud/density.pyx]: Add `xlabel="$r$", ylabel="$g(r)$"`

File: freud/density.pyx [hunk 22]
```diff
@@ -402,6 +408,12 @@ cdef class ComplexCF:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        return freud.common.line_plot(self.R, np.real(self.RDF),
+                                      title="Correlation Function",
+                                      xlabel=r"$r$",
+                                      ylabel=r"$\Re(C(r))$")
```
Comment (bdice) [inline freud/density.pyx]: Change to normal text Re

File: freud/density.pyx [hunk 23]
```diff
@@ -211,6 +211,12 @@ cdef class FloatCF:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        return freud.common.line_plot(self.R, self.RDF,
+                                      title="Correlation Function",
+                                      xlabel=r"$r$",
+                                      ylabel=r"$\langle p \ast q \rangle (r)$")
```
Comment (bdice) [inline freud/density.pyx]: Use `C(r)` instead of `p \ast q`

File: freud/common.py (hunk 20 above)
Comment (jinsooihm) [reply freud/common.py]: I made it to print `Keys of {} largest clusters (total clusters: {})`

Comment (jinsooihm) [reply freud/common.py]: Screen Shot 2019-06-10 at 3 02 54 PM
Screen Shot 2019-06-10 at 3 03 03 PM

File: freud/plot.py [hunk 24]
```diff
@@ -0,0 +1,272 @@
...
+                    [-2, -1, 0, 2], colors=colors)
+
+        fig.subplots_adjust(right=0.85)
+        cbar_ax = fig.add_axes([0.88, 0.1, 0.02, 0.8])
+        fig.colorbar(im, cax=cbar_ax)
+        f = io.BytesIO()
+        # Sets an Agg backend so this figure can be rendered
+        FigureCanvasAgg(fig)
+        fig.savefig(f, format='png')
+        return f.getvalue()
```
Comment (vyasr) [inline freud/plot.py]: A lot of this plotting logic is specific to the specific data in the PMFT example. I would simplify this to the bare minimum required to generate a simple PMFT plot without the bells and whistles to make the plot especially pretty. My inclination is to just do the simple heat map and remove the extra logic for the contour as well. @bdice may disagree with some of these recommendations though.

**Review by vyasr — COMMENTED** [review]
This looks really nice! I can't think of any other modules that should have png reprs added, @bdice can you?

File: freud/plot.py (hunk 24 above)
Comment (bdice) [reply freud/plot.py]: Yes, I agree we should simplify this to the minimum.

Comment (jinsooihm) [reply freud/plot.py]: Screen Shot 2019-06-13 at 9 40 19 PM

Something like this?
Or is this still specific to the data?

File: freud/environment.pyx [hunk 25]
```diff
@@ -869,6 +870,17 @@ cdef class MatchEnv:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        import plot
+        try:
+            count = np.unique(self.clusters, return_counts=True)
+        except ValueError:
+            return None
+        if not self.m_box.is2D():
```
Comment (bdice) [inline freud/environment.pyx]: Why do we limit to 2D if we're plotting `MatchEnv` cluster sizes?

File: freud/density.pyx [hunk 26]
```diff
@@ -402,6 +409,13 @@ cdef class ComplexCF:
     def __str__(self):
         return repr(self)
 
+    def _repr_png_(self):
+        import plot
+        return plot.line_plot(self.R, np.real(self.RDF),
+                              title="Correlation Function",
+                              xlabel=r"$r$",
+                              ylabel=r"$Re(C(r))$")
```
Comment (bdice) [inline freud/density.pyx]: Does this put `Re` in plain text or math font? I'm looking for this: 
image

File: freud/plot.py (hunk 24 above)
Comment (bdice) [reply freud/plot.py]: @jinsoo960 The code is doing a bunch of scaling and trickery that is specific to this hexagon analysis - the colorbar needs to find its min/max bounds from the min/max of the PMFT, and the x/y limits of the plot should correspond to the PMFT's actual bin locations. Last I checked, it was scaling the axes by 2/3 and only showing a cropped version of the actual data, for the purposes of this plot.

File: freud/density.pyx (hunk 26 above)
Comment (jinsooihm) [reply freud/density.pyx]: It seems like `matplotlib` does not have all the math fonts, but I think `operatorname{Re}` gets close to it.

File: freud/plot.py (hunk 24 above)
Comment (jinsooihm) [reply freud/plot.py]: Screen Shot 2019-06-13 at 10 46 06 PM
Screen Shot 2019-06-13 at 10 46 16 PM

This is an almost bare minimum version, with smoothing an all other logic removed.
Only processing done is changing `inf` to `nan`.

Comment (bdice) [reply freud/plot.py]: Close- bring back the part that labels the `x` and `y` ticks by the values of `pmft.X` and `pmft.Y`, instead of labeling them by the index of the bin.

Comment (bdice) [reply freud/plot.py]: Also change the title of the plot to `2D Potential of Mean Force`.

**Comment by bdice** [comment]
## To-do:
- [x] Take a look at pandas' [DataFrame.plot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.plot.html) - it takes `ax` as an argument and returns an `ax` object. We should refactor the `_repr_png_` methods into `plot` methods, so we can use this plotting code in our examples. There should be a separate `_repr_png_` method still, which will call `import plot; return plot.to_png_bytes(self.plot())` or something similar.
- [x] Add $k_B T$ to the colorbar of the PMFTXY2D, like this: 
image

**Review by bdice — APPROVED** [review]
Looks good to me!
